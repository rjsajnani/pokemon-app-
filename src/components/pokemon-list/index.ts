export * from './pokemon-list';
export * from './pokemon-list.props';
