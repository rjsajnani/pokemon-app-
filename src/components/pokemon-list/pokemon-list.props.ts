export type Pokemon = {
  url: string;
  id: number;
  name: string;
  image: string;
};

export type PokemonsListProps = {
  pokemons: Pokemon[];
  isNext: boolean;
  loadPokemons: () => void;
};
