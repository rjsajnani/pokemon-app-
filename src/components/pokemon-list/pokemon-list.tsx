import { ActivityIndicator, FlatList } from 'react-native';

import { TouchableOpacity } from 'react-native-gesture-handler';
import { useRouter } from 'expo-router';
import { Card } from '../card';
import useCapitalizedString from '../../hooks/useCapitalizedString';
import { Pokemon, PokemonsListProps as Props } from './pokemon-list.props';

const gap = 10;

export const PokemonList: React.FC<Props> = (props) => {
  const { pokemons, loadPokemons, isNext } = props;

  const router = useRouter();
  const capitalizedString = useCapitalizedString();

  const loadMore = () => {
    isNext && loadPokemons();
  };

  const renderItem = ({ item }: { item: Pokemon }) => {
    console.log(item, 'Item');
    return (
      <TouchableOpacity
        onPress={() => {
          router.push({
            pathname: 'details/id/',
            params: { id: '1', name: item.name },
          });
        }}
      >
        <Card
          image={item.image}
          id={item.id}
          name={capitalizedString(item.name)}
        />
      </TouchableOpacity>
    );
  };

  return (
    <FlatList
      renderItem={renderItem}
      data={pokemons}
      numColumns={2}
      keyExtractor={(item) => item.name}
      showsVerticalScrollIndicator={false}
      contentContainerStyle={{ gap }}
      columnWrapperStyle={{ gap }}
      onEndReached={loadMore}
      onEndReachedThreshold={0.1}
      ListFooterComponent={<ActivityIndicator size="large" color="#AEAEAE" />}
    />
  );
};
