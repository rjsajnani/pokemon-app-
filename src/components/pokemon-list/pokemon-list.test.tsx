import React from 'react';
import { render, RenderAPI } from '@testing-library/react-native';

import { PokemonList } from './pokemon-list';
import { Pokemon } from './pokemon-list.props';

const dummyPokemonList: Pokemon[] = [
  {
    url: 'pokemon1-url',
    id: 1,
    name: 'pokemon1',
    image: 'pokemonImage',
  },
  {
    url: 'pokemon2-url',
    id: 2,
    name: 'pokemon2',
    image: 'pokemonImage',
  },
  {
    url: 'pokemon3-url',
    id: 3,
    name: 'pokemon3',
    image: 'pokemonImage',
  },
];

describe('<PokemonList />', () => {
  let instance: RenderAPI;
  const loadPokemons = jest.fn();
  const nextUrl = false;

  beforeEach(function () {
    instance = render(
      <PokemonList
        pokemons={dummyPokemonList}
        loadPokemons={loadPokemons}
        isNext={dummyPokemonList.length > 0 ? false : !!nextUrl}
      />,
    );
  });

  it('renders correctly', () => {
    const tree = instance.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('check if the pokemon is renders or not ', async function () {
    expect(instance.getByText(/Pokemon3/i)).toBeTruthy();
  });
});
