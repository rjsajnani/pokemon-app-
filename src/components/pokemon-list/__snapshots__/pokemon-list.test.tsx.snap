// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonList /> renders correctly 1`] = `
<RCTScrollView
  ListFooterComponent={
    <ActivityIndicator
      color="#AEAEAE"
      size="large"
    />
  }
  contentContainerStyle={
    {
      "gap": 10,
    }
  }
  data={
    [
      {
        "id": 1,
        "image": "pokemonImage",
        "name": "pokemon1",
        "url": "pokemon1-url",
      },
      {
        "id": 2,
        "image": "pokemonImage",
        "name": "pokemon2",
        "url": "pokemon2-url",
      },
      {
        "id": 3,
        "image": "pokemonImage",
        "name": "pokemon3",
        "url": "pokemon3-url",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onEndReached={[MockFunction]}
  onEndReachedThreshold={0.1}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "gap": 10,
            },
          ]
        }
      >
        <RNGestureHandlerButton
          collapsable={false}
          delayLongPress={600}
          enabled={true}
          exclusive={true}
          handlerTag={1}
          handlerType="NativeViewGestureHandler"
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          touchSoundDisabled={false}
        >
          <View
            accessible={true}
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": "black",
                  "borderRadius": 10,
                  "height": 120,
                  "padding": 10,
                  "width": 337.5,
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "black",
                        "fontSize": 16,
                        "letterSpacing": 0.2,
                        "lineHeight": 24,
                      },
                      {
                        "color": "white",
                      },
                    ]
                  }
                >
                  Pokemon1
                </Text>
                <Text
                  style={
                    [
                      {
                        "color": "black",
                        "fontSize": 16,
                        "letterSpacing": 0.2,
                        "lineHeight": 24,
                      },
                      {
                        "color": "white",
                      },
                    ]
                  }
                >
                  #001
                </Text>
              </View>
              <Image
                source={
                  {
                    "uri": "pokemonImage",
                  }
                }
                style={
                  {
                    "bottom": 2,
                    "height": 80,
                    "position": "absolute",
                    "right": 2,
                    "width": 80,
                  }
                }
              />
            </View>
          </View>
        </RNGestureHandlerButton>
        <RNGestureHandlerButton
          collapsable={false}
          delayLongPress={600}
          enabled={true}
          exclusive={true}
          handlerTag={2}
          handlerType="NativeViewGestureHandler"
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          touchSoundDisabled={false}
        >
          <View
            accessible={true}
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": "black",
                  "borderRadius": 10,
                  "height": 120,
                  "padding": 10,
                  "width": 337.5,
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "black",
                        "fontSize": 16,
                        "letterSpacing": 0.2,
                        "lineHeight": 24,
                      },
                      {
                        "color": "white",
                      },
                    ]
                  }
                >
                  Pokemon2
                </Text>
                <Text
                  style={
                    [
                      {
                        "color": "black",
                        "fontSize": 16,
                        "letterSpacing": 0.2,
                        "lineHeight": 24,
                      },
                      {
                        "color": "white",
                      },
                    ]
                  }
                >
                  #002
                </Text>
              </View>
              <Image
                source={
                  {
                    "uri": "pokemonImage",
                  }
                }
                style={
                  {
                    "bottom": 2,
                    "height": 80,
                    "position": "absolute",
                    "right": 2,
                    "width": 80,
                  }
                }
              />
            </View>
          </View>
        </RNGestureHandlerButton>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "gap": 10,
            },
          ]
        }
      >
        <RNGestureHandlerButton
          collapsable={false}
          delayLongPress={600}
          enabled={true}
          exclusive={true}
          handlerTag={3}
          handlerType="NativeViewGestureHandler"
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          touchSoundDisabled={false}
        >
          <View
            accessible={true}
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                {
                  "backgroundColor": "black",
                  "borderRadius": 10,
                  "height": 120,
                  "padding": 10,
                  "width": 337.5,
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "black",
                        "fontSize": 16,
                        "letterSpacing": 0.2,
                        "lineHeight": 24,
                      },
                      {
                        "color": "white",
                      },
                    ]
                  }
                >
                  Pokemon3
                </Text>
                <Text
                  style={
                    [
                      {
                        "color": "black",
                        "fontSize": 16,
                        "letterSpacing": 0.2,
                        "lineHeight": 24,
                      },
                      {
                        "color": "white",
                      },
                    ]
                  }
                >
                  #003
                </Text>
              </View>
              <Image
                source={
                  {
                    "uri": "pokemonImage",
                  }
                }
                style={
                  {
                    "bottom": 2,
                    "height": 80,
                    "position": "absolute",
                    "right": 2,
                    "width": 80,
                  }
                }
              />
            </View>
          </View>
        </RNGestureHandlerButton>
      </View>
    </View>
    <View
      onLayout={[Function]}
    >
      <ActivityIndicator
        color="#AEAEAE"
        size="large"
      />
    </View>
  </View>
</RCTScrollView>
`;
